### ISSUE TITLE
How will Flutter implement Android 13's predictive back gesture?

### ISSUE THREAD

**Issue by victor-marino** [flutter/flutter#109513]
## Use case

According to [this page](https://developer.android.com/about/versions/13/features/predictive-back-gesture), we should get our apps ready for the new predictive back gesture that will be introduced in Android 13.

However, with just a few weeks to go before the final release, I haven't seen any info on how this will be handled in Flutter apps?

I have tried the simple trick of just adding `android:enableOnBackInvokedCallback="true"` to my Android Manifest, which does indeed activate the new animation on my Pixel 4 running the Android 13 beta. However, because all the back gestures are now captured by the system rather than Flutter, it completely breaks Flutter's navigation system, as ALL back gestures will now send you back to your phone's home screen, rather than taking you one level up Flutter's navigation tree when appropriate.

## Proposal

I'm merely asking for information here, as I've searched extensively but couldn't find anything pertaining Flutter.

Will we have to do anything to adapt our apps? Or will the Flutter team implement this at some point at framework level? If it's the latter, when can we expect this to be added? Android 13 is right around the corner.

I'm assuming this should be handled at framework level, so that Flutter automatically implements this animation when navigating inside's Flutter's own tree, and hands it over to the system when the user is already at the root of the navigation stack (so they can correctly peek at their home screen before completing the gesture).

Eager to get some info on this!

**Comment by timsneath** [comment]
Thanks so much for filing this issue! 

Just to clarify, this is a feature being introduced as part of a multi-year transition to how Android handles back gestures. While it's available to developers for the first time in Android 13 through a Developer Options setting, it's not currently accessible to users and won't be until a future Android release. So while the clock is indeed 'ticking' and we will work on enabling support for this, this is not something that Android 13 app authors need to worry about immediately.

**Comment by victor-marino** [comment]
Thanks a lot for your quick response! That clears things up for me.

I thought it would be available to regular users from the final release of Android 13 (without the need to flip the developer switch), but if that's not the case it's indeed not something to worry about for now.

**Comment by goderbauer** [comment]
My current understanding is that for this to work correctly, the framework will have to tell the engine (or android) up front whether it is going to handle the back gesture internally or whether Android should handle it.

@GaryQian I think you've been looking closer at the Android APIs. Does that match your understanding?

@chunhtai Does the Navigator already tell the engine that it can handle back internally? I seem to remember that we had to add this to support back on the web?

**Comment by chunhtai** [comment]
@goderbauer Currently flutter always tells Android it will handle the pop. If flutter then figure there is nothing else to pop, it then programmatically pop the android activity.

For web, the only thing I am aware is the fake browser entry to capture the backward button. I don't think it can be reuse here.

**Comment by goderbauer** [comment]
Thanks, so we will likely have to build something new here to tell the engine whether the framework can handle back or not upfront, if my understanding of the new Android API is correct.

**Comment by chunhtai** [comment]
cc @HansMuller @hangyujin 

I am still not sure whether we need to implement the animation. From the screenshot, it looks like this may only apply when popping the entire app. Is there any animation change when popping the in-app page?

**Comment by HansMuller** [comment]
There is, although support for that is farther out.

**Comment by GaryQian** [comment]
From looking at the API, it seems that we should always be telling Android that we handle the back gesture. Regardless of whether we implement predictive back or just legacy back on the framework side, we will be doing custom handling of the back gesture via OnBackPressedCallback (which is called by the OnBackPressedDispatcher).

However, we should add logic to let Android know when we run out of pages to pop. At this point, we should unregister the callback, and Android will then handle the next back as exiting the app.

In any case, the opt-in is in build.gradle and needs to be preconfigured. It shouldn't be possible (or necessary) for the app to dynamically inform the build whether to enable or disable predictive backnav if it is not already set before build.  We should be able to support this with a combo of Embedding API migration, framework predictive backnav implementation, and unregistering callbacks on final framework pop.

This is from an initial study of the API, I'll do some experimentation to see how it actually handles.

**Comment by victor-marino** [comment]
Hi all.

Just in case it's of any help, as I had already done some limited testing on my Pixel 4 running the Android 13 beta:

1. Even if I flip the developer switch for the predictive back gesture, my app won't show the gesture at all unless I have explicitly opted-in in the build.gradle, not even for the final gesture to go back to the homescreen. So unless something is done on Flutter/app side, the gesture will not appear at all.
2. If I opt-in in the build.gradle for my app, I do get the gesture working to back out of my app and into the homescreen. Bad news is, now EVERY back gesture takes me out of my app, even if there are still pages left to pop inside my Flutter app. I assume now every back gesture is being handled by the OS, so it just breaks Flutter's internal navigation.

So I think everything GaryQian said matches with my experience.

**Comment by stuartmorgan** [comment]
> 2\. now EVERY back gesture takes me out of my app, even if there are still pages left to pop inside my Flutter app. I assume now every back gesture is being handled by the OS

This will be fixed by https://github.com/flutter/flutter/issues/109558. Enabling the opt-in turns off the older back handling system.

**Comment by fullflash** [comment]
still back button detaches android app willpopscope not event catches back button event

Android 13 Samsung device.
latest flutter sdk

**Comment by deimantasa** [comment]
> still back button detaches android app willpopscope not event catches back button event
> 
> Android 13 Samsung device. latest flutter sdk

We had this issue as well. Few reasons it might happen:
1. Check your `AndroidManifest.xml` and remove if you've `android:enableOnBackInvokedCallback="true"`. We had it for some reason (I guess is a leftover from the past), which caused this issue
2. Check your `MainActivity` (java or kotlin). It might extend `FlutterFragmentActivity`.  If `android:enableOnBackInvokedCallback` is `true` in your manifest and `FlutterFragmentActivity` is extended, it seems to not work as well. However, what we've noticed that if `android:enableOnBackInvokedCallback` is set to `true` and `MainActivity` extends `FlutterActivity` - back button works as expected

Hope it helps.

**Comment by Akatsukishen** [comment]
@deimantasa You have done me a great favor. Thank you. When MainActivity extends FlutterFragmentActivity, the splash screen only show once, and it does not show when back to home and reopen app. Change MainActivity extends FlutterActivity, it shows every time.

**Comment by math1man** [comment]
FYI, https://github.com/flutter/engine/pull/35678 updated `FlutterActivity` but not `FlutterFragmentActivity`. That's why the latter doesn't work.

**Comment by RedsAnDev** [comment]
> FYI, [flutter/engine#35678](https://github.com/flutter/engine/pull/35678) updated `FlutterActivity` but not `FlutterFragmentActivity`. That's why the latter doesn't work.

Today this bug remains and all biometrics plugins cannot be used.

**Comment by justinmc** [comment]
Quick status update: This issue is actively being developed in https://github.com/flutter/flutter/pull/120385, which I hope to merge soon.  The engine PR https://github.com/flutter/engine/pull/39208 has already been merged.  Predictive back will then work at the root of Flutter apps built with the latest SDK on Android devices that support predictive back.

Later I also plan to implement predictive back route transitions that work with the native back arrow, but that will come in a separate PR.

**Comment by justinmc** [comment]
Predictive back now works when leaving Flutter apps.  Further work on predictive back route transitions will be tracked in https://github.com/flutter/flutter/issues/131961.

<img width="100" src="https://user-images.githubusercontent.com/389558/217918109-945febaa-9086-41cc-a476-1a189c7831d8.gif" />

## Getting predictive back to work in your Flutter app

  1. Run Android 33 or above.
  1. Enable the feature flag for predictive back on the device under "Developer
     options".
  1. Set `android:enableOnBackInvokedCallback="true"` in
     android/app/src/main/AndroidManifest.xml.
  1. Use the latest master branch of Flutter (or, later, make sure your version of Flutter is recent enough to include https://github.com/flutter/flutter/pull/120385).
  1. Run the app. Perform a back gesture (swipe from the left side of the
     screen).

You should see the predictive back animation and be able to commit or cancel it.

**Comment by milindgoel15** [comment]
Any issue created for this to be merged into a stable branch?

**Comment by justinmc** [comment]
This was reverted and I'm currently struggling to reland it here: https://github.com/flutter/flutter/pull/132249

Can you elaborate on the background problem?  If I open any Android app and immediately perform a back gesture, the app seems to be backgrounded and not killed.

**Comment by milindgoel15** [comment]
> If I open any Android app and immediately perform a back gesture, the app seems to be backgrounded and not killed.

Yes, so if predictive back gestures are set to false in the Android manifest, the app gets terminated. When reopening the app, it starts all the processes again. Check these recordings below.

This one doesn't have android:enableOnBackInvokedCallback="true" in the android manifest so the app gets terminated on using the back gesture:

https://github.com/flutter/flutter/assets/45682747/cc6ca334-94b8-4b91-9174-c86bc98ea032

This one has android:enableOnBackInvokedCallback="true" so the app appears in background mode so reopening the app launches in the last state:

https://github.com/flutter/flutter/assets/45682747/3473aac6-574c-45a8-97e4-d87326eb82e4


I feel it is the intended behaviour of using predictive back gestures because the demo provided in the [official page](https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture) also doesn't kill the application.

**Comment by justinmc** [comment]
@milindgoel15 I confirmed what you're saying by trying this on a native Android app with and without predictive back enabled, and then trying it on a Flutter app.  However, it seems to happen for me on master even though predictive back has been reverted, so I don't think it's directly related to PR https://github.com/flutter/flutter/pull/132249.

I've created a new issue for this: https://github.com/flutter/flutter/issues/132692.
